TypedSyntax/README.md

```diff
@@ -0,0 +1,156 @@
+
+The default is aimed at identifying sources of "type instability" (poor inferrability):
+
+```julia
+julia> printstyled(stdout, TypedSyntaxNode(f, (Float64, Int, Real)))
+```
+
+which produces
+
+<code>f(x, y, z::<b>Real</b>)::<b>Any</b> = (x + (y * z::<b>Real</b>)::<b>Any</b>)::<b>Any</b></code>
+
+The boldfaced text above is typically printed in color in the REPL:
+
+- red indicates non-concrete types
+- yellow indicates a "small union" of concrete types. These usually pose no issues, unless there are too many combinations of such unions.
+
+Printing with color can be suppressed with the keyword argument `iswarn=false`.
+
+## Caveats
+
+TypedSyntax aims for accuracy, but there are a number of factors that pose challenges.
+First, anonymous and internal functions appear as part of the source text, but internally Julia handles these as separate type-inferred methods, and these are hidden from the annotator.
+Therefore, in
+
+```julia
+julia> sumfirst(c) = sum(x -> first(x), c);    # better to use `sum(first, c)` but this is just an illustration
+
+julia> printstyled(stdout, TypedSyntaxNode(sumfirst, (Vector{Any},)))
+sumfirst(c)::Any = sum(x -> first(x), c)::Any
+```
+
+`x` and `first(x)` both have type `Any`, but they are not annotated as such because they are hidden inside the anonymous function.
+
+Second, this package works by attempting to "reconstruct history": starting from the type-inferred code, it tries to map calls back to the source. It would be much safer to instead keep track of the source during inference, but at present this is not possible (see [this Julia issue](https://github.com/JuliaLang/julia/issues/31162)). There are cases where this mapping fails: for example, with
+
+```julia
+julia> function summer(list)
+           s = 0
+           for x in list
+               s += x
+           end
+           return s
+       end;
+```
+the type-inferred code is (on Julia 1.9)
+```
+CodeInfo(
+    @ REPL[3]:2 within `summer`
+1 ─       (s = 0)::Core.Const(0)
+│   @ REPL[3]:3 within `summer`
+│   %2  = list::Vector{Float64}
+│         (@_3 = Base.iterate(%2))::Union{Nothing, Tuple{Float64, Int64}}
+│   %4  = (@_3 === nothing)::Bool
+│   %5  = Base.not_int(%4)::Bool
+└──       goto #4 if not %5
+2 ┄ %7  = @_3::Tuple{Float64, Int64}
+│         (x = Core.getfield(%7, 1))::Float64
+│   %9  = Core.getfield(%7, 2)::Int64
+│   @ REPL[3]:4 within `summer`
+│         (s = s + x)::Float64
+│   @ REPL[3]:5 within `summer`
+│         (@_3 = Base.iterate(%2, %9))::Union{Nothing, Tuple{Float64, Int64}}
+│   %12 = (@_3 === nothing)::Bool
+│   %13 = Base.not_int(%12)::Bool
+└──       goto #4 if not %13
+3 ─       goto #2
+    @ REPL[3]:6 within `summer`
+4 ┄       return s
+) => Union{Float64, Int64}
+```
+and very few calls here map to the source:
+```julia
+16-element Vector{Vector{Union{JuliaSyntax.TreeNode{JuliaSyntax.SyntaxData}, JuliaSyntax.TreeNode{TypedSyntax.TypedSyntaxData}}}}:
```

Comment:
How can we get this view?

Reply:
It required some internals. I refactored to make it easier to obtain and modified the README accordingly.